classes/local/tag/tag_manager.php

```diff
@@ -0,0 +1,183 @@
+     */
+    public const SYNC_STATUS_SYNC_NOT_REQUIRED = 1;
+
+    /**
+     * @var int Object tried to sync but there was an error. Will make it ignored and must be corrected manually.
+     */
+    public const SYNC_STATUS_ERROR = 2;
+
+    /**
+     * @var array All possible tag sync statuses.
+     */
+    public const SYNC_STATUSES = [
+        self::SYNC_STATUS_NEEDS_SYNC,
+        self::SYNC_STATUS_SYNC_NOT_REQUIRED,
+        self::SYNC_STATUS_ERROR,
+    ];
+
+    /**
+     * Returns an array of tag_source instances that are currently defined.
+     * @return array
+     */
+    public static function get_defined_tag_sources(): array {
+        // All possible tag sources should be defined here.
+        // Note this should be a maximum of 10 sources, as this is an AWS limit.
+        return [
+            new mime_type_source(),
+            new environment_source(),
+        ];
+    }
+
+    /**
+     * Is the tagging feature enabled and supported by the configured fs?
+     * @return bool
+     */
+    public static function is_tagging_enabled_and_supported(): bool {
+        $enabledinconfig = !empty(get_config('tool_objectfs', 'taggingenabled'));
+
+        $client = manager::get_client(manager::get_objectfs_config());
+        $supportedbyfs = !empty($client) && $client->supports_object_tagging();
+
+        return $enabledinconfig && $supportedbyfs;
+    }
+
+    /**
+     * Gathers the tag values for a given content hash
+     * @param string $contenthash
+     * @return array array of key=>value pairs, the tags for the given file.
+     */
+    public static function gather_object_tags_for_upload(string $contenthash): array {
+        $tags = [];
+        foreach (self::get_defined_tag_sources() as $source) {
+            $val = $source->get_value_for_contenthash($contenthash);
+
+            // Null means not set for this object.
+            if (is_null($val)) {
+                continue;
+            }
+
+            $tags[$source->get_identifier()] = $val;
+        }
+        return $tags;
+    }
+
+    /**
+     * Stores tag records for contenthash locally
+     * @param string $contenthash
+     * @param array $tags
+     */
+    public static function store_tags_locally(string $contenthash, array $tags) {
+        global $DB;
+
+        // Purge any existing tags for this object.
+        $DB->delete_records('tool_objectfs_object_tags', ['contenthash' => $contenthash]);
+
+        // Record time in var, so that they all have the same time.
+        $timemodified = time();
+
+        // Store new records.
+        $recordstostore = [];
+        foreach ($tags as $key => $value) {
+            $recordstostore[] = [
+                'contenthash' => $contenthash,
+                'tagkey' => $key,
+                'tagvalue' => $value,
+                'timemodified' => $timemodified,
+            ];
+        }
+        $DB->insert_records('tool_objectfs_object_tags', $recordstostore);
```

Comment:
Currently as it is setup all the tags for the same contenthash have the same timemodified. I left it this way because it is simpler to just add it here rather than what is probably more 'correct' which is having a single `timetagssynced` on the `tool_objectfs_objectfs` record that is updated.

I think I might remove the timemodified here and add that in, unless we decide / work out if we are going to diff them or just reset the state each time (above comment)

Reply:
I've changed it over now so the single object record has the `tagpushedtime`, instead of each tag record individually. I don't think there is enough value in having each have a timemodified to justify the required extra complexity of calculating and storing the diff (maybe in the future as a separate pr)